# Лабораторная работа № 4. Нейронные сети

Постройте нейросетевой классификатор для набора данных [MNIST](https://www.kaggle.com/c/digit-recognizer).

## Задание

1. Постройте двуслойную нейронную сеть.
2. Постройте отчет по классификации  [classification_report](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html) и матрицу ошибок [confusion_matrix](http://scikit-learn.org/stable/auto_examples/model_selection/plot_confusion_matrix.html#sphx-glr-auto-examples-model-selection-plot-confusion-matrix-py)
3. Добавьте скрытые слои и оцените как изменится качество классификации.

## Начало работы

In [66]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from tensorflow import keras
from tensorflow.keras.layers import Dense

In [55]:
from keras.models import Model
from keras.layers import Input
visible = Input(shape=(784,), name ='Input_784')
hidden = Dense(300, activation = 'relu', name ='Dense_300')(visible)
output = Dense(10, activation = 'softmax', name ='Dense_10')(hidden)
model_1 = Model(inputs=visible, outputs=output)
model_1.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 Input_784 (InputLayer)      [(None, 784)]             0         
                                                                 
 Dense_300 (Dense)           (None, 300)               235500    
                                                                 
 Dense_10 (Dense)            (None, 10)                3010      
                                                                 
Total params: 238,510
Trainable params: 238,510
Non-trainable params: 0
_________________________________________________________________


In [56]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [57]:
from keras.datasets import mnist
from keras import utils
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [59]:
from tensorflow.keras.utils import to_categorical

In [61]:
x_train = x_train.astype('float32')
x_train /= 255
x_train = x_train.reshape(-1, 784)
x_test = x_test.astype('float32')
x_test /= 255
x_test = x_test.reshape(-1, 784)
y_train_ohe =to_categorical(y_train, 10)
y_test_ohe = to_categorical(y_test, 10)
print(y_train[0], y_train_ohe[0])

5 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]


In [64]:
model.fit(x_train, y_train_ohe, batch_size=128, epochs=12, verbose=1, validation_split=0.2)

Epoch 1/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2997 - accuracy: 0.1145 - val_loss: 2.3007 - val_accuracy: 0.1060
Epoch 2/12
375/375 [==============================] - 1s 3ms/step - loss: 2.3000 - accuracy: 0.1156 - val_loss: 2.3001 - val_accuracy: 0.1060
Epoch 3/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2999 - accuracy: 0.1146 - val_loss: 2.3002 - val_accuracy: 0.1060
Epoch 4/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2997 - accuracy: 0.1147 - val_loss: 2.3006 - val_accuracy: 0.1035
Epoch 5/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2995 - accuracy: 0.1177 - val_loss: 2.3013 - val_accuracy: 0.1060
Epoch 6/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2996 - accuracy: 0.1134 - val_loss: 2.3022 - val_accuracy: 0.1060
Epoch 7/12
375/375 [==============================] - 1s 3ms/step - loss: 2.2994 - accuracy: 0.1173 - val_loss: 2.3002 - val_accuracy: 0.1060
Epoch 

Точность

In [65]:
accuracy = model.evaluate(x_test, y_test_ohe)
print("Loss:", accuracy[0])
print("Accuracy:", accuracy[1])

313/313 [==============================] - 0s 1ms/step - loss: 2.2989 - accuracy: 0.1135
Loss: 2.2989439964294434
Accuracy: 0.11349999904632568


В сравнении с другими методами обучения, система обучения нейронной сети показала результат в 11%. Что на 70% меньше, чем метод соседей или логистической регрессии. 
Тест проведен на GOOGLE COLAB, который не использует GPU. Хотя скорость прохождения одной эпохи в 1 секунду не очень плохая.
 